File: C#/Chess.cs
/*******************************************************************************
 * N queens problem (C# version)
 * Class Chess
 *
 * Copyleft 2013 Vikman - All rights revoked.
 * vikman90.blogspot.com - vmfdez90@gmail.com
 * February 8, 2013
 * 
 ******************************************************************************/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Queens
{
    class Chess
    {
        //----------------------------------------------------------------------
        // Constructor

        public Chess(int size)
        {
            byte[] completeSet = Enumerable.Repeat((byte)1, size).ToArray();

            this.size = size;
            nSteps = 0;
            nDiscards = 0;
            queens = new byte[size][];
            queensCount = Enumerable.Repeat(size, size).ToArray();

            for (int i = 0; i < size; i++)
                this.queens[i] = (byte[])completeSet.Clone();

            discardedPairs = new Stack<int[]>();
            discardedCount = new Stack<int>();
            random = new Random();
        }

        //----------------------------------------------------------------------
        // Solve queens problem

        public bool Solve()
        {
            int index = SelectIndex();
            int nvalues;
            byte[] currentSet;
            int[] values = new int[size];

            if (index == -1)
                return true;

            currentSet = (byte[])queens[index].Clone();
            nvalues = SelectValues(index, values);

            for (int i = 0; i < nvalues; i++) 
            {
                int value = values[i];

                if (!Assign(index, value))
                {
                    queens[index] = (byte[])currentSet.Clone();
                    queensCount[index] = nvalues;
                    continue;
                }

                if (Solve())
                    return true;

                RestoreLast();
                queens[index] = (byte[])currentSet.Clone();
                queensCount[index] = nvalues;
            }

            return false;
        }

        //----------------------------------------------------------------------
        // Representation of the chessboard

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder();

            for (int i = 0; i < size; i++)
                if (queensCount[i] != 1)
                    builder.Append("Reina " + (i + 1) + " no resuelta.\n");
                else
                    builder.Append("Reina " + (i + 1) + ": casilla " + (Array.IndexOf(queens[i], (byte)1) + 1) + "\n");

            return builder.ToString();
        }

        //----------------------------------------------------------------------
        // Get number of total tries of assignation

        public long Steps
        {
            get
            {
                return nSteps;
            }
        }

        //----------------------------------------------------------------------
        // Get number of total discards

        public long Discards
        {
            get
            {
                return nDiscards;
            }
        }

        //----------------------------------------------------------------------
        // Assign a value to a row and propagate constraints

        private bool Assign(int index, int value)
        {
            Array.Clear(queens[index], 0, size);
            queensCount[index] = 0;
            discardedCount.Push(0);
            nSteps++;

            for (int i = 0; i < size; i++)
            {
                if (i == index)
                    continue;

                int diag1 = value + (index - i);
                int diag2 = value - (index - i);

                if (!(Discard(i, value) && Discard(i, diag1) && Discard(i, diag2)))
                {
                    RestoreLast();
                    return false;
                }
            }

            queens[index][value] = 1;
            queensCount[index] = 1;
            return true;
        }

        //----------------------------------------------------------------------
        // Discard candidate values (constraints propagation)

        private bool Discard(int index, int value)
        {
            if (value < 0 || value >= size || queens[index][value] == 0)
                return true;

            nDiscards++;
            queens[index][value] = 0;
            queensCount[index]--;
            discardedPairs.Push(new int[] {index, value});
            discardedCount.Push(discardedCount.Pop() + 1);

            if (queensCount[index] == 0)
                return false;

            if (queensCount[index] == 1)
            {
                value = Array.IndexOf(queens[index], (byte)1);

                for (int i = 0; i < size; i++)
                {
                    if (i == index)
                        continue;

                    int diag1 = value + (index - i);
                    int diag2 = value - (index - i);

                    if (!(Discard(i, value) && Discard(i, diag1) && Discard(i, diag2)))
                        return false;
                }
            }

            return true;
        }

        //----------------------------------------------------------------------
        // Undo last assignation (restore constraints)

        private void RestoreLast()
        {
            int n = discardedCount.Pop();

            for (int i = 0; i < n; i++)
            {
                int[] pair = discardedPairs.Pop();
                
                if (queens[pair[0]][pair[1]] == 0)
                {
                    queens[pair[0]][pair[1]] = 1;
                    queensCount[pair[0]]++;
                }
            }
        }

        //----------------------------------------------------------------------
        // Get index of a unsolved row (minimum remaining values heuristics)

        private int SelectIndex()
        {
            int minCount = size + 1;
            int index = -1;

            for (int i = 0; i < size; i++)
            {
                if (queensCount[i] > 1 && queensCount[i] < minCount)
                {
                    index = i;
                    minCount = queensCount[i];
                }
            }

            return index;
        }

        //----------------------------------------------------------------------
        // Select all available indices from a row

        private int SelectValues(int index, int[] values)
        {
            int nvalues = 0;
            int offset = random.Next(size);

            for (int i = 0; i < size; i++)
            {
                int value = (offset + i) % size;

                if (queens[index][value] != 0)
                    values[nvalues++] = value;
            }

            return nvalues;
        }

        //----------------------------------------------------------------------

        readonly int size;                      // Number of queens
        private long nSteps;                    // Number of calls to Assign()
        private long nDiscards;                 // Number of calls to Discard()
        private byte[][] queens;                // Queens' positions (set of candidate positions)
        private int[] queensCount;              // Number of available values
        private Stack<int[]> discardedPairs;    // Discarded candidates (index-value)
        private Stack<int> discardedCount;      // Number of discards in the last assignation
        private Random random;                  // Generator of random numbers
    }
}

File: PHP/chess.php
<?php
/*******************************************************************************
 * N queens problem (PHP version)
 * Class Chess
 *
 * Copyleft 2016 Vikman - All rights revoked.
 * vikman90.blogspot.com - vmfdez90@gmail.com
 * March 24, 2016
 *
 ******************************************************************************/

mt_srand();

class Chess {
    private $size;
    private $queens;
    private $nsteps;
    private $stackDiscarded;
    private $stackCount;

    //--------------------------------------------------------------------------
    // Constructor

    public function __construct($size) {
        $this->size = $size;
        $this->queens = array_fill(0, $size, array_fill(0, $size, 1));
        $this->nsteps = 0;
        $this->stackDiscarded = [];
        $this->stackCount = [];
    }

    //--------------------------------------------------------------------------
    // Solve queens problem

    public function solve() {
        $index = $this->selectIndex();

        if ($index == -1)
            return true;

        $row = $this->queens[$index];
        $values = $this->selectValues($row);

        foreach ($values as $value) {
            if (!$this->assign($index, $value)) {
                $this->queens[$index] = $row;
                continue;
            }

            if ($this->solve())
                return true;

            $this->restoreLast();
            $this->queens[$index] = $row;
        }

        return false;
    }

    //--------------------------------------------------------------------------
    // Get index of a unsolved row (minimum remaining values heuristics)

    private function selectIndex() {
        $minSize = $this->size + 1;
        $index = -1;

        for ($i = 0; $i < $this->size; $i++) {
            $curSize = count($this->queens[$i]);

            if ($curSize > 1 and $curSize < $minSize) {
                $index = $i;
                $minSize = $curSize;
            }
        }

        return $index;
    }

    //--------------------------------------------------------------------------
    // Assign a value to a row and propagate constraints

    private function assign($index, $value) {
        $this->queens[$index] = [];
        $this->stackCount[] = 0;
        $this->nsteps++;

        for ($i = 0; $i < $this->size; $i++) {
            if ($i == $index)
                continue;

            $diag1 = $value + ($index - $i);
            $diag2 = $value - ($index - $i);

            if (!($this->discard($i, $value) and $this->discard($i, $diag1) and
                  $this->discard($i, $diag2))) {
                $this->restoreLast();
                return false;
            }
        }

        $this->queens[$index][$value] = 1;
        return true;
    }

    //--------------------------------------------------------------------------
    // Discard candidate values (constraints propagation)

    private function discard($index, $value) {
        $row = $this->queens[$index];

        if (!isset($row[$value]))
            return true;

        unset($row[$value]);
        $this->stackDiscarded[] = [$index, $value];
        $this->stackCount[count($this->stackCount) - 1]++;

        if (count($row) == 0)
            return false;

        if (count($row) == 1) {
            $value = array_keys($row)[0];

            for ($i = 0; $i < $this->size; $i++) {
                if ($i == $index)
                    continue;

                $diag1 = $value + ($index - $i);
                $diag2 = $value - ($index - $i);

                if (!($this->discard($i, $value) and $this->discard($i, $diag1) and
                      $this->discard($i, $diag2))) {
                    $this->restoreLast();
                    return false;
                }
            }
        }

        return true;
    }

    //--------------------------------------------------------------------------
    // Undo last assignation (restore constraints)

    private function restoreLast() {
        $n = array_pop($this->stackCount);

        for ($i = 0; $i < $n; $i++) {
            $pair = array_pop($this->stackDiscarded);
            $this->queens[$pair[0]][$pair[1]] = 1;
        }
    }

    //--------------------------------------------------------------------------
    // Get number of total tries of assignation

    private function selectValues($array) {
        $values = [];
        $offset = mt_rand(0, $this->size - 1);

        for ($i = 0; $i < count($array); $i++) {
            $index = ($offset + $i) % $this->size;

            if (isset($array[$index]))
                $values[] = $index;
        }

        return $values;
    }

    //--------------------------------------------------------------------------
    // Get number of total tries of assignation

    public function steps() {
        return $this->nsteps;
    }

    //--------------------------------------------------------------------------
    // Representation of the chessboard

    public function __toString() {
        $string = "";

        for ($i = 0; $i < $this->size; $i++) {
            $j = $i + 1;
            $value = array_keys($this->queens[$i])[0] + 1;
            $string .= "Reina $j: $value\n";
        }

        return $string;
    }
}
